File: qbopt/abi/callsite.py
from dataclasses import replace

from iced_x86 import Code
from iced_x86 import Register

from qbopt.abi import runtime
from qbopt.frontend import blocks
from qbopt.frontend.declen import Insn
from qbopt.objectfile.module import Module


def caller_cleanup(following: Insn | None) -> bool:
    if following is None:
        return False
    instruction = following.insn
    return (
        instruction.code in (Code.ADD_RM16_IMM8, Code.ADD_RM16_IMM16)
        and instruction.op0_register == Register.SP
        and 0 < instruction.immediate(1) < 0x8000
    )


def inferred(found: Module, contracts: dict[int, runtime.Contract]) -> dict[int, runtime.Contract]:
    candidates = {
        at: rule for at, rule in contracts.items() if rule.inputs is None and not rule.name.upper().startswith("B$")
    }
    if not candidates:
        return {}
    decoded = blocks.instructions(found)
    if isinstance(decoded, str):
        return {}
    by_address = {one.at: one for one in decoded}
    inputs = frozenset({runtime.Reg.AX, runtime.Reg.BX, runtime.Reg.CX, runtime.Reg.DX, runtime.Reg.SI, runtime.Reg.DI})
    result = {}
    for at, rule in candidates.items():
        call = by_address.get(at)
        if call is None:
            continue
        caller = caller_cleanup(by_address.get(call.end))
        result[at] = replace(
            rule,
            inputs=inputs,
            cleanup=0 if caller else rule.cleanup,
            evidence=(
                "C caller cleanup assumed from adjacent ADD SP; "
                if caller
                else "Pascal callee cleanup assumed; argument byte count remains unknown; "
            )
            + "language ABI takes no incoming arithmetic flags; all GP inputs retained; "
            "memory, clobber and control effects unchanged",
        )
    return result

File: src/abi/callsite.rs
//! Port of `qbopt/abi/callsite.py`: a language ABI's inputs and cleanup for
//! a call the runtime table does not establish.

use std::collections::BTreeSet;

use iced_x86::{Code, Register};

use crate::abi::runtime::{Contract, Reg};
use crate::frontend::blocks;
use crate::frontend::declen::Insn;
use crate::objectfile::module::Module;
use crate::support::hash::IndexMap;

pub fn caller_cleanup(following: Option<&Insn>) -> bool {
    let Some(following) = following else {
        return false;
    };
    let instruction = &following.insn;
    matches!(instruction.code(), Code::Add_rm16_imm8 | Code::Add_rm16_imm16)
        && instruction.op0_register() == Register::SP
        && 0 < instruction.immediate(1)
        && instruction.immediate(1) < 0x8000
}

pub fn inferred(found: &Module, contracts: &IndexMap<i64, Contract>) -> IndexMap<i64, Contract> {
    let candidates: IndexMap<i64, &Contract> = contracts
        .iter()
        .filter(|(_, rule)| rule.inputs.is_none() && !rule.name.to_uppercase().starts_with("B$"))
        .map(|(&at, rule)| (at, rule))
        .collect();
    if candidates.is_empty() {
        return IndexMap::default();
    }
    let Ok(decoded) = blocks::instructions(found) else {
        return IndexMap::default();
    };
    let by_address: IndexMap<usize, &Insn> = decoded.iter().map(|one| (one.at, one)).collect();
    let inputs: BTreeSet<Reg> = BTreeSet::from([Reg::Ax, Reg::Bx, Reg::Cx, Reg::Dx, Reg::Si, Reg::Di]);
    let mut result = IndexMap::default();
    for (&at, rule) in &candidates {
        let Some(call) = by_address.get(&(at as usize)) else {
            continue;
        };
        let caller = caller_cleanup(by_address.get(&call.end()).copied());
        result.insert(
            at,
            Contract {
                inputs: Some(inputs.clone()),
                cleanup: if caller { Some(0) } else { rule.cleanup },
                evidence: format!(
                    "{}language ABI takes no incoming arithmetic flags; all GP inputs retained; \
                     memory, clobber and control effects unchanged",
                    if caller {
                        "C caller cleanup assumed from adjacent ADD SP; "
                    } else {
                        "Pascal callee cleanup assumed; argument byte count remains unknown; "
                    }
                ),
                ..(*rule).clone()
            },
        );
    }
    result
}

#[cfg(test)]
#[path = "callsite_tests.rs"]
mod tests;
